### PR TITLE
feat: add indexing methods

### PR DESCRIPTION
**Pull request by averikitsch**
Instead of https://github.com/googleapis/langchain-google-cloud-sql-pg-python/pull/12

### CODE REVIEW

File: src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py [hunk 1]
```diff
@@ -290,3 +297,53 @@ def similarity_search(
...
+        filter = f"WHERE ({index.partial_indexes})" if index.partial_indexes else ""
+        params = "WITH " + index.index_options()
+        concurrently = "CONCURRENTLY" if concurrently else ""
+
+        if index.distance_strategy == DistanceStrategy.EUCLIDEAN:
+            function = "vector_l2_ops"
+        elif index.distance_strategy == DistanceStrategy.COSINE_DISTANCE:
+            function = "vector_cosine_ops"
+        else:
+            function = "vector_ip_ops"
```
Comment (kurtisvg) [inline src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py:319]: I saw a bunch of similiar logic in the other PR -- would it be better to abstract into the DistanceStrategy class?

File: src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py [hunk 2]
```diff
@@ -290,3 +297,53 @@ def similarity_search(
...
+
+        name = name if name else index.name
```
Comment (kurtisvg) [inline src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py:321]: nit: you can also use "something = x or y" which seems a bit more readable

File: src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py [hunk 3]
```diff
@@ -290,3 +297,53 @@ def similarity_search(
         self, query: str, k: int = 4, **kwargs: Any
     ) -> List[Document]:
         return []
+
+    async def aapply_vector_index(
+        self,
+        index: BaseIndex,
+        name: Optional[str] = None,
```
Comment (kurtisvg) [inline src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py:304]: Should this allow a Vector object as well?

File: src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py [hunk 4]
```diff
@@ -290,3 +297,53 @@ def similarity_search(
...
+            function = "vector_l2_ops"
+        elif index.distance_strategy == DistanceStrategy.COSINE_DISTANCE:
+            function = "vector_cosine_ops"
+        else:
+            function = "vector_ip_ops"
+
+        name = name if name else index.name
+        # Concurrently errors
+        stmt = f"CREATE INDEX  {name} ON {self.table_name} USING {index.index_type} ({self.embedding_column} {function}) {params} {filter};"
+        print(stmt)
```
Comment (kurtisvg) [inline src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py:324]: Should these prints be removed?

**Review by kurtisvg — CHANGES_REQUESTED** [review]

File: src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py (hunk 3 above)
Comment (averikitsch) [reply src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py:304]: Can you add more? Since I have this on the vector store I don't think so?

Comment (kurtisvg) [reply src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py:421]: should be able to do this: (might need to use `Union[str, BaseIndex]` depends on when the other format was added python version wise) 
```suggestion
        name: Optional[str | BaseIndex] = None,
```

File: src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py [hunk 5]
```diff
@@ -407,3 +414,47 @@ def similarity_search(
         self, query: str, k: int = 4, **kwargs: Any
     ) -> List[Document]:
         return []
+
+    async def aapply_vector_index(
+        self,
+        index: BaseIndex,
+        name: Optional[str] = None,
+        concurrently=False,
```
Comment (kurtisvg) [inline src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py:422]: nit: type hint missing

File: src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py [hunk 6]
```diff
@@ -407,3 +414,47 @@ def similarity_search(
...
+    ) -> None:
+        if isinstance(index, ExactNearestNeighbor):
+            return None
```
Comment (kurtisvg) [inline src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py:425]: nit:
```suggestion
            return
```

File: src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py [hunk 7]
```diff
@@ -407,3 +414,47 @@ def similarity_search(
...
+
+        filter = f"WHERE ({index.partial_indexes})" if index.partial_indexes else ""
```
Comment (kurtisvg) [inline src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py:427]: nit: could do
```suggestion
        filter = f"WHERE ({index.partial_indexes or ""})"
```

File: src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py (hunk 6 above)
Comment (kurtisvg) [inline src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py:425]: Dumb question -- should we delete an index with the name if it already exists?

File: src/langchain_google_cloud_sql_pg/indexes.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+
+@dataclass
+class BaseIndex(ABC):
+    name: str = DEFAULT_INDEX_NAME
+    index_type: str = "base"
+    distance_strategy: DistanceStrategy = field(
+        default_factory=lambda: DistanceStrategy.COSINE_DISTANCE
+    )
+    partial_indexes: Optional[List] = None
```
Comment (kurtisvg) [inline src/langchain_google_cloud_sql_pg/indexes.py:47]: is this: 
```suggestion
    partial_indexes: Optional[List[str]] = None
```

**Review by kurtisvg — APPROVED** [review]

File: src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py (hunk 3 above)
Comment (averikitsch) [reply src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py:726]: We are already requiring a BaseIndex as the index argument, why does name need to support BaseIndex as well?

File: src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py (hunk 6 above)
Comment (averikitsch) [reply src/langchain_google_cloud_sql_pg/cloudsql_vectorstore.py]: Not dumb! We did talk about this and I agree we should drop

**Pull request merged**